### PR TITLE
adding F-contig layout support + other cleanups

### LINKED ISSUE #89: [BUG] `matmul` produces incorrect results for F-contiguous / non-row-major inputs
`np.matmul` on QuadPrecision arrays returns incorrect results whenever either input is F-contiguous (or otherwise not row-major-contiguous in its last two axes).

## Reproducer

<details>
<summary>click to expand</summary>

```python
import numpy as np
from numpy_quaddtype import QuadPrecision, QuadPrecDType

dt = QuadPrecDType(backend='sleef')
def Q(v): return QuadPrecision(str(float(v)), backend='sleef')

A = np.asfortranarray(np.array([[Q(1), Q(2), Q(3)],
                                [Q(4), Q(5), Q(6)]], dtype=dt))
B = np.asfortranarray(np.array([[Q(1), Q(2)],
                                [Q(3), Q(4)],
                                [Q(5), Q(6)]], dtype=dt))

# A is F-contig: strides=(16, 32); B is F-contig: strides=(16, 48)
print("quad F@F:", [[float(v) for v in r] for r in (A @ B)])
# -> [[23.0, 27.0], [35.0, 32.0]]   <- WRONG

# Same call against native dtypes works fine — proving the bug is on our side:
A32 = np.asfortranarray(np.array([[1., 2., 3.], [4., 5., 6.]], dtype=np.float32))
B32 = np.asfortranarray(np.array([[1., 2.], [3., 4.], [5., 6.]], dtype=np.float32))
print("float32 F@F:", (A32 @ B32).tolist())
# -> [[22.0, 28.0], [49.0, 64.0]]   <- correct

# Workaround: force C-contig before matmul
print("quad (C@C):",
      [[float(v) for v in r] for r in (np.ascontiguousarray(A) @ np.ascontiguousarray(B))])
# -> [[22.0, 28.0], [49.0, 64.0]]   <- correct
````

</details>

## Root cause

`quad_matmul_strided_loop_aligned` in `src/csrc/umath/matmul.cpp` unconditionally:

```cpp
size_t lda = A_row_stride / sizeof(Sleef_quad);
// ...
qblas_gemm('R', 'N', 'N', m, p, n, &alpha, A_ptr, lda, B_ptr, ldb,
           &beta, C_ptr, ldc_numpy);
```

This silently assumes:

- the input is **row-major contiguous** in its core dimensions and the right layout argument to QBLAS is always `'R'`.

## Possible fixes
- Inspect each operand's last-two-axis strides: dispatch directly to `qblas_gemm('R', ...)` or `('C', ...)` when both operands share a contiguous layout, otherwise memcpy the offending operand(s) into a row-major temp buffer (same trick the unaligned path already uses) and call `qblas_gemm('R', ...)`.
- **Performance (needs QBLAS-side change too).** For mixed-layout inputs (R @ C and C @ R) the copy above can be skipped by passing `transa`/`transb='T'` instead, which would require `qblas_gemm` to gain transpose support. See https://github.com/SwayamInSync/QBLAS/issues/13

## Related tests

On the #88 two tests already exercise this bug and are marked `xfail`:

- `tests/test_dot.py::TestMatmulBatched::test_batched_fortran_order`
- `tests/test_dot.py::TestMatmulBatched::test_batched_transposed_view`

### PR DESCRIPTION
**Pull request by SwayamInSync**
closes #89 

This PR improves the BLAS integration very parallel to the NumPy's own logic.

- Mirrors NumPy's own `@TYPE@_matmul` dispatch. NumPy hands the inner loop raw strides and expects the loop itself to dispatch, so we do the same instead of assuming row-major.
- Ports NumPy's `is_blasable2d` check: an operand is BLAS-able iff its fast axis is contiguous and the slow axis is a valid leading dimension; each operand is tested in both orientations (C- and F-blasable).
- BLAS-able operands are passed to QBLAS zero-copy with the matching transpose flag (`'N'/'T'`) and `lda` derived from strides, so F-contiguous/transposed inputs need no copy (QBLAS already supports `transa/transb` correctly).
- Non-BLAS-able operands (fully strided, negative strides) and non-row-major output are gathered into contiguous temps and scattered back, exactly NumPy's copy fallback.
- Oversized (`dim > INT_MAX`) or empty cores fall back to the naive strided loop, matching NumPy's `too_big_for_blas` guard.

I think this is a good base to integrate dot operations efficiently  in follow-up PRs

### CODE REVIEW

**Comment by SwayamInSync** [comment]
Ahh the fix to run older-cpu workflows is done in #104 , once that will merge we can update here.

File: src/csrc/umath/matmul.cpp [hunk 1]
```diff
@@ -126,192 +187,223 @@ quad_matmul_strided_loop_aligned(PyArrayMethod_Context *context, char *const dat
...
     return 0;
 }
 
+static int
+quad_matmul_strided_loop_unaligned(PyArrayMethod_Context *context, char *const data[],
+                                   npy_intp const dimensions[], npy_intp const strides[],
+                                   NpyAuxData *auxdata)
+{
+    return naive_matmul_strided_loop(context, data, dimensions, strides, auxdata);
+}
```
[the remark this answers is not in the document]
Comment (SwayamInSync) [reply src/csrc/umath/matmul.cpp:405]: numpy's allocator returns ≥16-aligned memory, so ordinary quad arrays are aligned

**Comment by SwayamInSync** [comment]
cc: @ngoldbaum this one too :)

File: src/csrc/quadblas_interface.cpp [hunk 2]
```diff
@@ -46,6 +46,19 @@ qblas_gemv(char layout, char trans, size_t m, size_t n,
...
+            return -1;
+        }
+        Sleef_quad zero = Sleef_cast_from_doubleq1(0.0);
+        int beta_zero = Sleef_icmpeqq1(*beta, zero);
+        for (size_t i = 0; i < y_len; i++) {
+            y[i * incy] = beta_zero ? zero : Sleef_mulq1_u05(*beta, y[i * incy]);
+        }
         return 0;
     }
     if (!alpha || !A || !x || !beta || !y) {
```
Comment (ngoldbaum) [inline src/csrc/quadblas_interface.cpp]: You already checked `!beta` above so might as well delete that here

File: tests/test_dot.py [hunk 3]
```diff
@@ -1054,3 +1044,119 @@ def test_batched_larger(self, batch, m, k, n):
...
+    the naive loop. Every case is cross-checked against float64."""
+
+    @staticmethod
+    def _layout(a, kind):
+        if kind == "C":
+            return np.ascontiguousarray(a)
+        if kind == "F":
+            return np.asfortranarray(a)
+        if kind == "T":
+            return np.ascontiguousarray(a.T).T
```
Comment (ngoldbaum) [inline tests/test_dot.py]: Isn't this the same as F-order? Maybe makes more sense to just return `a.T`, which is different from a fortran-order copy.

File: src/csrc/quadblas_interface.cpp [hunk 4]
```diff
@@ -66,7 +79,7 @@ qblas_gemm(char layout, char transa, char transb,
            Sleef_quad *B, size_t ldb,
            Sleef_quad *beta, Sleef_quad *C, size_t ldc)
 {
-    if (m == 0 || n == 0 || k == 0) {
+    if (m == 0 || n == 0) {
```
Comment (ngoldbaum) [inline src/csrc/quadblas_interface.cpp:82]: Claude says that no tests hit this code change. You could add a test that does `np.matmul(a, b)` where the contracted dimension is zero to hit this.

File: tests/test_dot.py [hunk 5]
```diff
@@ -1054,3 +1044,119 @@ def test_batched_larger(self, batch, m, k, n):
...
+        A_f = rng.standard_normal((20, 16))
+        B_f = rng.standard_normal((16, 24))
+        _assert_matmul_matches_float64(_qnd(A_f)[::2, ::2], _qnd(B_f)[::2, ::2],
+                                       A_f[::2, ::2], B_f[::2, ::2])
+
+    def test_negative_stride_input(self):
+        rng = np.random.default_rng(seed=201)
+        A_f = rng.standard_normal((6, 5))
+        B_f = rng.standard_normal((5, 7))
+        _assert_matmul_matches_float64(_qnd(A_f)[::-1], _qnd(B_f), A_f[::-1], B_f)
```
Comment (ngoldbaum) [inline tests/test_dot.py:1156]: Maybe add a test for negative column strides too? e.g. `A[:, ::-1]`.

File: tests/test_dot.py [hunk 6]
```diff
@@ -1054,3 +1044,119 @@ def test_batched_larger(self, batch, m, k, n):
...
+
+    def test_strided_A_fortran_B_fortran_out_three_temps(self):
+        rng = np.random.default_rng(seed=202)
+        A_f = rng.standard_normal((20, 7))
+        B_f = rng.standard_normal((7, 11))
+        out = np.asfortranarray(np.zeros((10, 11), dtype=QuadPrecDType()))
```
Comment (ngoldbaum) [inline tests/test_dot.py:1169]: Maybe also add a test for strided `out`? Something like `out=np.empty(14, dtype=QuadPrecDType())[::2]; np.matmul(A, B, out=out)`.

**Review by ngoldbaum — APPROVED** [review]
Overall looks good, just a few minor suggestions inline.

**Comment by SwayamInSync** [comment]
Thanks @ngoldbaum I applied the reviews and extra tests here and it is now ready to merge!

**Comment by SwayamInSync** [comment]
failures seems independent, not quite sure why `apt` its unreachable, chatgpt is recommending to keep trying in a loop :)

**Comment by ngoldbaum** [comment]
Thanks!
